File: pset6/credit/credit.py
# Reverse number given
card = input("Number: ")[::-1]
sum = 0
count = 0


def main():
    # If valid, state card type
    if CheckValidity():
        PrintCards()
    # Else print INVALID
    else:
        print("INVALID")


def CheckValidity():
    # Go through each letter
    for digit in card:
        # Increment counter
        global count
        global sum

        count += 1
        # If counter is even
        if count % 2 == 0:
            # Multiply the current digit by two
            newNo = int(digit) * 2
            # Get the first and second digit if number is greater than 9 and add to sum
            if newNo > 9:
                sum += newNo % 10
                sum += newNo // 10
            # Get digit if less than 9 and add to sum
            else:
                sum += newNo
        # If counter is odd, add to sum variable
        else:
            sum += int(digit)
    return (sum % 10 == 0)


# Determine the type of card
def PrintCards():
    # AMEX
    if len(card) == 15 and card[count - 1] == "3" and (card[count - 2] == "4" or card[count - 2] == "7"):
        print("AMEX")
    # MASTERCARD
    elif len(card) == 16 and card[count - 1] == "5" and int(card[count - 2]) in range(1, 6):
        print("MASTERCARD")
    # VISA
    elif (len(card) == 13 or len(card) == 16) and card[count - 1] == "4":
        print("VISA")
    # INVALID
    else:
        print("INVALID")


main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

bool CheckSum(long no);
void CardType(void);

long digits[16];
int counter = 0;

int main(void)
{
    long cardNo = get_long("Number: ");
    if (CheckSum(cardNo))
    {
        CardType();
    }

    else
    {
        printf("INVALID\n");
    }
}

bool CheckSum(long no)
{
    int sum = 0;

    // Save each digit to an array
    while (no)
    {
        digits[counter++] = no % 10;
        no /= 10;
    }

    // Add every other number starting from the second to last digit
    for (int i = 1; i < counter; i += 2)
    {
        long val = digits[i] * 2;
        // If the value has two digits and the first and last digit seperately
        if (val > 9)
        {
            sum += val % 10;
            sum += val / 10;
        }

        else
        {
            sum += val;
        }
    }

    // Add numbers not previously added
    for (int i = 0; i < counter; i += 2)
    {
        sum += digits[i];
    }

    // If the last digit is 0, return true
    if (sum % 10 == 0)
    {
        return true;
    }

    return false;
}

void CardType()
{
    // VISA
    if ((counter == 13 || counter == 16) && digits[counter - 1] == 4)
    {
        printf("VISA\n");
    }

    // AMEX
    else if (counter == 15 && digits[counter - 1] == 3 && (digits[counter - 2] == 7 || digits[counter - 2] == 4))
    {
        printf("AMEX\n");
    }

    // MASTERCARD
    else if (counter == 16 && digits[counter - 1] == 5 && (digits[counter - 2] >= 1 && digits[counter - 2] <= 5))
    {
        printf("MASTERCARD\n");
    }

    else
    {
        printf("INVALID\n");
    }
}
